### ISSUE TITLE
Best way to handle sub-queries

### ISSUE THREAD

**Issue by kunalgrover05** [noripyt/django-cachalot#108]
Thank you for the excellent documentation written in your project, it gave me an idea of different caching libraries, and how they differ.

I have a use-case which might not really be common and I couldn't find anything relevant in docs, so I wanted to discuss it here. I am trying to understand what is the best way to handle sub-queries. (The documentation mentioned that cachalot handles sub-queries, but I couldn't find any in-depth information of whether the below case is handled)

My site has very few visitors, and both consistent Reads/Writes are important so I have never thought about adding Caching.
- Now we are working to develop some tools which allow us to view some aggregate data from the site. These are not just aggregate SQL queries, but also joined data from different models, eg: Users with their profiles and all related information in one screen.  (These queries may be read inconsistent and there are no writes)
- When the number of users grow a lot, these JOINs become very expensive(we need to access all data without pagination)
- On top of these, from the frontend, there are filters which allow users to view only some part of the data which means that using View or Query based Caches are not going to work very well.

I have the following thoughts ahead:
- Cache the whole queryset manually(including prefetches). All filters are run using custom code in the whole queryset.
- Create new tables containing JOINed data. The data will still be served from DB but much faster without JOINs. On top of this, 

What is your opinion here? I am completely off here?

**Comment by BertrandBordage** [comment]
What I meant by “django-cachalot handles subqueries” is simply that they are correctly taken into account when caching and invalidating data. This means a query containing any number of ultra-complex non-raw subqueries will still be cached, and it will correctly be invalidated if any of the tables from that query (and its subqueries) gets modified.

I see your point, I work on a similar project, [dezede/dezede](https://github.com/dezede/dezede), that also displays a ton of related data using a lot of `JOIN`s.

What I suggest you to do is to simply try using django-cachalot in development and test it against the most expensive queries. Pay attention to the RAM usage, it may be too big for your server.
Without django-debug-toolbar (it has a huge overhead on SQL queries), test the time spent on these pages with and without django-cachalot.

If the benefits are negligible, don’t use django-cachalot. Your best option will be to code optimisations and systems such as an infinite pagination.

In the future, please ask questions [on gitter](https://gitter.im/django-cachalot/Lobby).

**Comment by kunalgrover05** [comment]
Thank you for your response. I felt a Github issue is much better since others might benefit by searching.

I am sorry if I framed my question incorrectly, I will try to repeat it in a simpler way. I have only one queryset with lot of filters applied on top. eg:
```
User.objects.select_related('profile').all()
User.objects.select_related('profile').all(Q(profile__location='A')|Q(name='A'))
User.objects.select_related('profile').all(Q(profile__location='B'))
User.objects.select_related('profile').all(Q(profile__age__gt=19))
```
The dimensionality of queries is very high but the number of queries are very low.

Is there any way to optimize these kind of queries?
(Which is why I suggested the mechanism of keeping the whole User queryset in cache and writing a custom script to do filtering for me.)

**Comment by kunalgrover05** [comment]
Based on further analysis I did, I would like to answer the above question. The answer is that **it doesn't matter**.
A caching library does not need to try to keep becoming smarter (it is not competing to be better than the database). I would like to answer some observations from further analysis done by me.

- I analysed a lot of queries done from my DB, and the problem always is a lot of SQL queries. I analyzed those queries using the helpful django-debug-toolbar and silk
I was able to reduce time for executing SQL queries using prefetching. (I brought down 2000 SQL queries to 38)

- Django caching API allows to take care of simple business cases. I had a use-case of reading nearly all data from DB. Since this was aggregate data and consistency doesn't matter, I simply cached each object(including relationships) from the DB manually. (and obtained qs from DB still) 
This simply worked, and invalidation are done based on default settings.

Why the optimizations were unnecessary
--
If you don't have a lot of time to spend on the above ideas, there's a way which is 1000 times simpler. Simply adding cachalot to my application allowed me to get enough gains that I wouldn't have needed to do any further deep-dives above.

**Why does Cachalot work in the case above?**
Because that's how we usually code with the ORM. We ask for relationships when we want them, and not doing pre-fetches/optimizations helps us to write code which is easily cachable.

I will give a real world example here from the LeanAgri setup:

```python
class Crop(Model)
    
class Fertilizer(Model)

class Schedule(Model)
    crop = ForeignKey(Crop)
    fertilizer = ManyToManyField(Fertilizer)

for schedule in Schedule.objects.all():
    print schedule.crop.get_data()
    for fertilizer in schedule.fertilizer.all():
          print fertilizer.get_data()
```

Why was it written this way? Because that's how a lot of code was written to begin with. There were no thoughts of optimizations, there were business use-cases which keep on getting added over time. 
**Cachalot will infact perform better** without these optimizations added since multiple objects will share the same queries and thus be able to take advantage of those cachings.

However, not optimising our code to use pre-fetches limits our un-cached queries and make them a lot slower. So it is a good idea to still optimize and pre-fetch. 

**TLDR**: It doesn't matter what you "think" would your use case be and whether it would be handled well by the caching setup. To put it simply, it will be. Unless you dig deep into what is actually happening in your application, the no-brainer way of putting cachalot in your application is going to make things considerably better.

**Comment by BertrandBordage** [comment]
Glad to see django-cachalot worked particularly well in your project :)
